lib/AST/GenericSignatureBuilder.cpp

```diff
@@ -5073,31 +5073,47 @@ class GenericSignatureBuilder::InferRequirementsWalker : public TypeWalker {
     }
 
     // SWIFT_ENABLE_TENSORFLOW
-    // Infer `Differentiable` or `Differentiable & AdditiveArithmetic` generic
-    // constraints from `@differentiable` or `@differentiable(linear)`.
+    // Infer requirements from `@differentiable` or `@differentiable(linear)`
+    // function types.
+    // For all non-`@noDerivative` parameter and result types:
+    // - `@differentiable`: add `T: Differentiable` requirement.
+    // - `@differentiable(linear)`: add
+    //   `T: Differentiable`, `T == T.TangentVector` requirements.
     if (auto *fnTy = ty->getAs<AnyFunctionType>()) {
       if (fnTy->isDifferentiable()) {
-        auto addConstraint = [&](Type typeToConstrain, ProtocolDecl *protocol) {
-          Requirement req(RequirementKind::Conformance, typeToConstrain,
+        auto addConformanceConstraint = [&](Type type, ProtocolDecl *protocol) {
+          Requirement req(RequirementKind::Conformance, type,
                           protocol->getDeclaredType());
           Builder.addRequirement(req, source, nullptr);
         };
-        auto constrainParametersAndResult = [&](ProtocolDecl *protocol) {
+        auto addSameTypeConstraint = [&](Type firstType,
+                                         AssociatedTypeDecl *assocType) {
+          auto *protocol = assocType->getProtocol();
+          auto conf = Builder.lookupConformance(CanType(), firstType, protocol);
+          auto secondType = conf.getAssociatedType(
+              firstType, assocType->getDeclaredInterfaceType());
+          Requirement req(RequirementKind::SameType, firstType, secondType);
+          Builder.addRequirement(req, source, nullptr);
+        };
+        auto &ctx = Builder.getASTContext();
+        auto *differentiableProtocol =
+            ctx.getProtocol(KnownProtocolKind::Differentiable);
+        auto *tangentVectorAssocType =
+            differentiableProtocol->getAssociatedType(ctx.Id_TangentVector);
+        auto addRequirements = [&](Type type, bool isLinear) {
+          addConformanceConstraint(type, differentiableProtocol);
+          if (isLinear)
+            addSameTypeConstraint(type, tangentVectorAssocType);
```

Comment:
If you add `T.TangentVector == Self`, then you don't need to add `T: AdditiveArithmetic` anymore I believe.

Reply:
`T: AdditiveArithmetic` isn't added anymore.